joomla/joomla-cms
# [5.3] SEF: Fix URLs when preprocessing

### Summary of Changes
The routing in Joomla requires some information to be in the query to build proper URLs. One is the `slug`, the combination of the ID and alias of a content item (for example `&id=42:the-answer-to-everything`) and the other is the parent key, if the extension requires this. (for example `&catid=21`) When this information is missing, the router can't build the right URL.

This PR adds a new component router rule, which should help fixing such URLs. The rule is supposed to be flexible enough to be used by most third party components. The parameters for the rule are the view configuration to act upon, the table to read the info from, the tables key and the tables parent key. The parent key is optional and the rule can be added more than once for different views.


### Testing Instructions
1. You need at least one article which is not directly linked to by a menu item.
2. Edit an article and insert a link like `index.php?option=com_content&view=article&id=<id>` where `id` is the ID of the article which is NOT directly linked to by a menu item.
3. Check out the link in the frontend by visiting the article.


### Actual result BEFORE applying this Pull Request
You get a strange link, for example something like `/menuitem?view=article&id=42`


### Expected result AFTER applying this Pull Request
You get the correct link, for example `/menuitem/this-test-is-successfull`


### Link to documentations
Please select:
